gql/dsl.py

```diff
@@ -151,17 +152,37 @@ def serialize_list(serializer, list_values):
     return ListValueNode(values=FrozenList(serializer(v) for v in list_values))
 
 
+class SerializerWrapper:
+    def __init__(self):
+        self._serializer = None
+
+    def set_serializer(self, serializer):
+        self._serializer = serializer
+        return self._serializer
+
+    def __call__(self, *args, **kwargs):
+        return self._serializer(*args, **kwargs)
+
+
+visited_arg_types: Dict[Any, SerializerWrapper] = {}
+
+
 def get_arg_serializer(arg_type):
     if isinstance(arg_type, GraphQLNonNull):
         return get_arg_serializer(arg_type.of_type)
     if isinstance(arg_type, GraphQLInputField):
         return get_arg_serializer(arg_type.type)
     if isinstance(arg_type, GraphQLInputObjectType):
+        if arg_type in visited_arg_types:
```

Comment:
I admit that I don't really understand all the dsl code completely, but is the SerializerWrapper class really necessary ?
Isn't it equivalent to:

```python
    if isinstance(arg_type, GraphQLInputObjectType):                                                                    
        if arg_type in visited_arg_types:                                                                               
            return visited_arg_types[arg_type]                                                                          
                                                                                                                        
        visited_arg_types[arg_type] = None                                                                              
                                                                                                                        
        serializers = {k: get_arg_serializer(v) for k, v in arg_type.fields.items()}                                    
                                                                                                                        
        visited_arg_types[arg_type] = lambda value: ObjectValueNode(                                                    
            fields=FrozenList(                                                                                          
                ObjectFieldNode(name=NameNode(value=k), value=serializers[k](v))                                        
                for k, v in value.items()                                                                               
            )                                                                                                           
        )                                                                                                               
                                                                                                                        
        return visited_arg_types[arg_type]
```

Reply:
The wrapper is necessary. It sets an anchor while there is no actual serializer's body finalized.
Your code wouldn't work as the infinite recursion happens at `serializers = {k: get_arg_serializer(v) for k, v in arg_type.fields.items()}`, it will never get a chance to hit `visited_arg_types[arg_type] = lambda value: ObjectValueNode(`, thus `if arg_type in visited_arg_types:` will always be `False`.

Reply:
Ignore my previous comment. I think you are right.